Updated the script to return the original position/rotation

TheYellowArchitect opened the pull request:
- `induce_stress` now allows for one-liner to start shaking, instead of 2 lines (shaker.target = someNode3D)
- `shake_once` was deleted because it's obsolete.
- The demo was updated for it as well.

TheYellowArchitect committed:
Updated the script.

Dragon20C commented:
Thank you for making a pull request, I will take a look, though I wont be able to push it today, so tomorrow I will do a thorough look.

Dragon20C commented:
@TheYellowArchitect Hello I have taken a look and I see some improvements that I already pushed to the main repo, please take a look, anyways you have changed some things I would not agree with.


1. Removing the shake_once (now renamed to shake_node) was implemented like this because currently godot 4.3 does not have a way to export a button so this is a work around for creating a button, now why this exists is because I want to have the ability for to see the shake in the editor, this avoids going back and forth into the game.

2. You have added some checks for example checking for null in the target, that makes sense to me so I have added that bit to the main repo, thank you!

3. You wanted to have the ability to shake multiple objects at once but currently it wont work because of the way it shakes one node(try to add another cube and use the same nodeshaker variable), I have a rough idea on fixing this so multiple objects can be shaken. 

4. In the ready function we don't need to set the initial position and rotation as the setting of the target actually does this (this was from previous old code, now removed and fixed)

5. I think you have over-complicated the resting the initial position & rotation, I have implemented something simple, though I do understand why you have it this way, if you want to have multiple objects shake, this makes sense to me. 

Thank you for writing a pr and hope I have changed somethings to your liking and about shaking multiple objects at once I think I can implement it, if you have some ideas please ask here!

TheYellowArchitect commented:
>3. You wanted to have the ability to shake multiple objects at once but currently it wont work because of the way it shakes one node(try to add another cube and use the same nodeshaker variable), I have a rough idea on fixing this so multiple objects can be shaken.

I didn't add shaking multiple objects at once. I only added the ability to reset the currently shaking node, if another node is added to shake.

This feature would be quite big imo, and would require basically an array of "shakers", each tracking the object they shake, the current trauma etc

>4. In the ready function we don't need to set the initial position and rotation as the setting of the target actually does this (this was from previous old code, now removed and fixed)

I wasn't certain if a target being set from the editor gets into the setter, so I let it in to not break anything :sweat_smile: 

>5. I think you have over-complicated the resting the initial position & rotation, I have implemented something simple, though I do understand why you have it this way, if you want to have multiple objects shake, this makes sense to me.

This is my use-case scenario. Think any card game like slay the spire or whatever. I have 1 nodeshaker in the scene, and many cards in the player's hand. If a player clicks a disabled card, the card shakes (negative visual feedback)
So the nodeshaker starts without a target. If a player clicks a disabled card, it sets the new target. But if player re-clicks a card which is already shaking, it should take that in mind, hence the additional check. If a different card is clicked, it resets the previous target as seen.

>Thank you for writing a pr

np, I did these changes to make it work for my project, so I had to share to help others.
 **Freely edit this PR to your liking :)**

>and about shaking multiple objects at once I think I can implement it

I think it would complicate the code. The simplicity of this plugin is its greatest strength, just 70 lines in 1 script, for amazing functionality. The competitor "Shaker" has shit tons of scripts for advanced functionality which is bloat if you don't need that.

As a result, I would rather just copy-paste nodeshaker nodes, because this works good for a single target. Expanding it to multiple targets at once via arrays, is easy but would also make it bloated. But then again, it would require making it a singleton, so might as well make it a seperate plugin lol
I may do it in the future in this project I'm working on though. If I do, I will push a PR.

Thinking about it further, seeing that you plan for a nodeshaker to support exclusively 1 target, the nodeshaker should be a child node of the target node imo. That's intuitive to me.

Dragon20C commented:
@TheYellowArchitect Thank you for replying, I am glad someone found a use for it and has given me valuable feedback, I think the plugin is in a better state now, I might make a branch that includes the ability to shake multiple objects at once, Im feeling up for the challenge, since I feel like the pr is done I am closing it, please feel free to open up an issue here if you have other ideas.

Dragon20C commented:
@TheYellowArchitect Hello, So I have pushed something to the multi_node_shaker branch (https://github.com/Dragon20C/Godot-NodeShaker3D/tree/Shake_Multiple_Nodes) its not final as I think you had a good way of structuring it so I will steal some of your ideas (yoink!) so feel free to test this branch and give me feedback on it, it should work and all the shakes for each node3d is different.

TheYellowArchitect commented:
I apologize for not replying earlier, I just saw my email messages. I have been busy, but I will try to review it in the next few days (I hope the multi-node-shaker is a singleton, if not that's my first feedback :P )
